File: system.go
package main

// SystemAdapter abstracts all external command execution and OS interactions.
// Real implementations call wg, ipset, and ip commands.
// Fake implementations are used in tests.
type SystemAdapter interface {
	// IsRoot returns true when the process is running as root (uid 0).
	IsRoot() bool

	// InterfaceSubnet returns the CIDR subnet of the given network interface,
	// e.g. "10.8.0.0/24".
	InterfaceSubnet(iface string) (string, error)

	// WGDump runs "wg show <iface> dump" and returns the raw output.
	WGDump(iface string) (string, error)

	// IPSetList runs "ipset list <setname> -o save" and returns the raw output.
	IPSetList(setname string) (string, error)

	// IPSetCreate creates an ipset with the given type and family inet.
	// withComment enables per-entry comment storage on the set.
	// The creation is idempotent: it is equivalent to "ipset create ... -exist".
	IPSetCreate(setname, setType string, withComment bool) error

	// IPSetAdd runs "ipset add <setname> <entry> [comment <comment>]".
	IPSetAdd(setname, entry, comment string) error

	// IPSetDel runs "ipset del <setname> <entry>".
	IPSetDel(setname, entry string) error

	// WGSetPeer adds or updates a WireGuard peer on the interface.
	WGSetPeer(iface, pubkey, allowedIP string) error

	// WGDelPeer removes a WireGuard peer from the interface.
	WGDelPeer(iface, pubkey string) error

	// WGGenKey generates a new WireGuard private key string.
	WGGenKey() (string, error)

	// WGPubKey derives the public key from a private key string.
	WGPubKey(privkey string) (string, error)
}

File: rust-port/src/system.rs
pub const DEFAULT_CONFIG_DIR: &str = "/etc/wireguard/wgman";

pub trait SystemAdapter {
    fn is_root(&self) -> bool;
    fn stdout_is_terminal(&self) -> bool;
    fn stderr_is_terminal(&self) -> bool;
    fn interface_subnet(&self, iface: &str) -> Result<String, String>;
    fn wg_dump(&self, iface: &str) -> Result<String, String>;
    fn ipset_list(&self, set_name: &str) -> Result<String, String>;
    fn ipset_create(
        &self,
        set_name: &str,
        set_type: &str,
        with_comment: bool,
    ) -> Result<(), String>;
    fn ipset_add(&self, set_name: &str, entry: &str, comment: &str) -> Result<(), String>;
    fn ipset_del(&self, set_name: &str, entry: &str) -> Result<(), String>;
    fn wg_set_peer(&self, iface: &str, pub_key: &str, allowed_ip: &str) -> Result<(), String>;
    fn wg_del_peer(&self, iface: &str, pub_key: &str) -> Result<(), String>;
    fn wg_gen_key(&self) -> Result<String, String>;
    fn wg_pub_key(&self, private_key: &str) -> Result<String, String>;
}
